fix: GAD source version parsing (#2793)

b31ngd3v opened the pull request:
part of #2793

b31ngd3v commented:
the tests here will fail because it's using cached database so it already has the wrong version parsed, after it gets merged the next time when the database gets cached it'll parse the right version.
https://github.com/b31ngd3v/cve-bin-tool/actions/runs/4391627473 passed all the test on my local branch!
![image](https://user-images.githubusercontent.com/64733912/224477672-4ec40e1e-e91f-4653-b6a8-6c00977abd2d.png)

anthonyharrison reviewed (changes requested):
@b31ngd3v I don't think you can simply remove the ) and ] characters from the version string. The ) means the version up to excluding the value; the ] means the version  up to and including the value

b31ngd3v commented:
@anthonyharrison i didn't replace the characters only, first it checks if the version string contains any of that character if yes, then it does the appropriate thing (as you said). please check the screenshot below.

![image](https://user-images.githubusercontent.com/64733912/224687599-3b69546a-d310-4584-8a90-a195c9226d75.png)

terriko commented on cve_bin_tool/data_sources/gad_source.py:
```diff
@@ -189,14 +189,30 @@ def parse_range_string(self, range_string):
             versions = version_string.replace(",", " ").split(" ")
 
             for version in versions:
+                if (
+                    version.replace(")", "")
+                    .replace("(", "")
+                    .replace("[", "")
+                    .replace("]", "")
+                    == ""
+                ):
```
Ugh, this works but it's kind of confusing to do actual string replaces here so it probably needs a comment or possibly needs to be written in a different way?  

I *think* what you're trying to do is make sure there's at least one character in here that isn't a bracket.  Since a regex with a bunch of brackets in it is just as hard to read as this is (`"[[\]()]"`), you could maybe go the other way and look for an actual version number to improve readability:

```suggestion
                # Make sure we have an actual version number and not just a bunch of brackets
                if not re.search("[0-9]", version):
```

This will fail if we have versions that are only letters, but I'm pretty sure those don't work right now for other reasons.

BTW, you probably can't just accept this suggestion without making changes because I'm not sure there's an `import re` line above. And you don't need to accept my suggestion if you just put an explanation in a comment so it's clear what you're doing.

b31ngd3v replied:
thanks @terriko 
this seems like the best way to do this.

terriko commented:
Do we also have to have some logic for the VersionStartExcluding and VersionStartIncluding options here?

b31ngd3v commented:
> Do we also have to have some logic for the VersionStartExcluding and VersionStartIncluding options here?

i've updated the logic, can you please re-verify the code? thanks @terriko

anthonyharrison commented:
> Do we also have to have some logic for the VersionStartExcluding and VersionStartIncluding options here?

@b31ngd3v @terriko For Java, see [here](https://maven.apache.org/pom.html#Dependency_Version_Requirement_Specificationurl)

The syntax for versions is dependent on the package manager see the [affected range note](https://gitlab.com/gitlab-org/security-products/gemnasium-db/-/blob/master/README.md) for more details

b31ngd3v commented:
thanks @anthonyharrison this helps a lot, i'll modify the code accordingly. thank you again.

b31ngd3v commented:
@anthonyharrison @terriko all the cases are taken care of except the last two, how do we handle those? also version strings like `affected_range: "[4.4.0],(,4.2.0)"`

should we split them in two records?

![image](https://user-images.githubusercontent.com/64733912/225394615-2ed0c582-568b-495e-b58b-a3cbe2c0430f.png)

anthonyharrison commented:
> @anthonyharrison @terriko all the cases are taken care of except the last two, how do we handle those? also version strings like `affected_range: "[4.4.0],(,4.2.0)"`
> 
> should we split them in two records?
> 
> ![image](https://user-images.githubusercontent.com/64733912/225394615-2ed0c582-568b-495e-b58b-a3cbe2c0430f.png)

Thanks @b31ngd3v for the analysis. I would suggest that we create two records for the cases which aren't covered. @terriko  do you agree?

b31ngd3v commented:
done! fixed version parsing and tests also!! @terriko @anthonyharrison requesting review

terriko commented:
Windows tests timed out entirely so I'm going to give re-running them a shot.

terriko commented:
Hey, what's the status on this one?  It looks like there are tests failing but it doesn't look related.

b31ngd3v commented:
@terriko everything is done, also added tests! the tests here will fail because it's using cached database so it already has the wrong version parsed, after it gets merged the next time when the database gets cached it'll parse the right version.
https://github.com/b31ngd3v/cve-bin-tool/actions/runs/4552300734 passed all the test on my local branch (because i deleted the cached db before running the tests)!
![image](https://user-images.githubusercontent.com/64733912/228561613-5cb8b56a-8d8c-4b0e-928b-1b1980612b5b.png)

terriko reviewed (approved):
Okay, then I'm going to mark this as approved, but I might not come back to actually merge it for a couple of hours so I can limit the amount of time I spend dealing with the same "broken until cache update" issue on main.  Thanks so much for your work on this!

terriko commented:
Merging now before I switch to other tasks this afternoon.  I expect this will break tests on main for around 5 hours until the cache is updated, so I warned folk on gitter.
